lib/redis/distributed.rb

```diff
@@ -799,32 +803,57 @@ def punsubscribe(*channels)
     end
 
     # Watch the given keys to determine execution of the MULTI/EXEC block.
-    def watch(*_keys)
-      raise CannotDistribute, :watch
+    def watch(*keys, &block)
+      ensure_same_node(:watch, keys) do |node|
+        @watch_key = key_tag(keys.first) || keys.first.to_s
+
+        begin
+          node.watch(*keys, &block)
+        rescue StandardError
+          @watch_key = nil
```

Comment:
That's not good enough, you want to reset `@watch_key` from an `ensure` block. Here you only reset it on error.

Reply:
Hum, actually scratch that, it seems `watch do` is not expected to unwatch... that's a very weird behavior :/